docs/RFCS/20220120_row_level_ttl.md

```diff
@@ -0,0 +1,364 @@
+process which does not need heavy monitoring.
+
+### Rate Limiting
+In an ideal world, admission control could reject traffic and force a backoff
+and retry if it detects heavy load and rate limiting is not needed. However,
+we will introduce a rate limiter option just in case admission control is not
+adequate, especially in its infancy of being exposed. We can revisit this at a
+later point.
+
+### Controlling Delete Rate
+Whilst our aim is for foreground traffic to be unaffected whilst deletion rate
+manages to clean up "enough", we expect that there is still tinkering developers
+must make to ensure a rate for their workload.
+
+As such, users have knobs that control the delete rate, including:
+* how often the deletion job runs (controls amount of "junk" data left)
+* table GC time (when tombstones are removed and space is therefore reclaimed)
+* the size of the ranges on the table, which has knock on effects for
+  `ttl_range_concurrency`.
+
+As part of the `(option = value, …)` syntax, users can also control the
+following on the deletion job itself:
+
+Option | Description
+--- | ---
+`expire_after` | When a TTL would expire. Accepts any interval. Defaults to ''30 days''. Minimum of `'5 minutes'`.
+`ttl_select_batch_size` | How many rows to fetch from the range that have expired at a given time. Defaults to 500. Must be at least `1`.
+`ttl_delete_batch_size` | How many rows to delete at a time. Defaults to 100. Must be at least `1`.
+`ttl_range_concurrency` | How many concurrent ranges are being worked on at a time. Defaults to `cpu_core_count`. Must be at least `1`.
+`ttl_admission_control_priority` | Priority of the admission control to use when deleting. Set to a high amount if deletion traffic is the priority.
+`ttl_select_as_of_system_time` | AS OF SYSTEM TIME to use for the SELECT clause. Defaults to `-30s`. Proposed as a hidden option, as I'm not sure anyone will actually need it.
+`ttl_maximum_rows_deleted_per_second` | Maximum number of rows to be deleted per second (acts as the rate limit). Defaults to 0 (signifying none).
+`ttl_pause` | Pauses the TTL job from executing. Existing jobs will need to be paused using CANCEL JOB.
+`ttl_job_cron` | Frequency the job runs, specified using the CRON syntax.
+
+#### A note on delete_batch_size
+Deleting rows on tables with secondary indexes on a table lays more intents,
+hence causing higher contention. In this case, there is an argument to delete
+less rows at a time (or indeed, just one row at a time!) to avoid this. However,
+testing showed deletes could never catch up in this case and realistically a
+very small number of rows can be deleted in this way. "100" was set as the
+balance for now.
+
+## Filtering out Expired Rows
+Rows that have expired their TTL can be optionally removed from all SQL query
+results. However, this is not planned to be implemented for the first release of
+TTL.
+
+#### User override for the `crdb_internal_expiration` column
```

Comment:
The subheading level here is kind of strange, made me think that this was a subsection of `Filtering out Expired Rows`